Remove unnecessary py::object copy in PyRRef ctor

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38415 Enforce const on PyRRef functions
* **#38402 Remove unnecessary py::object copy in PyRRef ctor**
* #38376 Use GIL to guard decref of jit::toPyObj return value in processRpc
* #38348 Use GIL to guard py::object's destruction
* #38366 Explicitly decref py::object in PythonRpcHandler
* #38364 Explicitly decref py::object in ConcretePyObjectHolder and PythonFunctionGuard
* #38340 Minor code cleanup

Differential Revision: [D21554724](https://our.internmc.facebook.com/intern/diff/D21554724)

Commit:
Remove unnecessary py::object copy in PyRRef ctor

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -108,8 +108,10 @@ PyRRef::PyRRef(const py::object& value, const py::object& type_hint)
+        // jit::toIValue takes a py::handle as the first argument, and it calls
+        // py::handle.cast<py::object>() to incref of provided value. The
+        // returned ivalue will keep the reference alive.
+        IValue ivalue = jit::toIValue(value, elem_type);
```
I think this simplification makes sense.

There is one subtle change here. 

If we have this `py::object` copy, the passed-in `const py::object&` must be kept alive until `py::object copy(value);` to make `PyRRef::PyRRef` work correctly.

If we remove this `py::object` copy, the passed-in `const py::object&` must be kept alive until the `py::handle.cast<py::object>()` in  `jit::toIValue(..)`.

So let's mention this lifetime assumption on the passed-in `const py::object&`, and why we need this the pybind11 binding for this constructor to be holding GIL.

mrshenli replied:
comments added

Commit:
Update on "Remove unnecessary py::object copy in PyRRef ctor"




[ghstack-poisoned]

Commit:
Update on "Remove unnecessary py::object copy in PyRRef ctor"


Differential Revision: [D21554724](https://our.internmc.facebook.com/intern/diff/D21554724)

[ghstack-poisoned]

Commit:
Update on "Remove unnecessary py::object copy in PyRRef ctor"


Differential Revision: [D21554724](https://our.internmc.facebook.com/intern/diff/D21554724)

[ghstack-poisoned]

Commit:
Update on "Remove unnecessary py::object copy in PyRRef ctor"


Differential Revision: [D21554724](https://our.internmc.facebook.com/intern/diff/D21554724)

[ghstack-poisoned]
